RISCV: Remove shouldForceRelocation and unneeded relocations

MaskRay opened the pull request:
Follow-up to #140494

`shouldForceRelocation` is conservative and produces redundant
relocations.

For example, RISCVAsmBackend::ForceRelocs (introduced to support mixed
relax/norelax code) leads to redundant relocations in the following
example adapted from #77436

```
.option norelax
j label
// For assembly input, RISCVAsmParser::ParseInstruction sets ForceRelocs (https://reviews.llvm.org/D46423).
// For direct object emission, RISCVELFStreamer sets ForceRelocs (#77436)
.option relax
call foo  // linker-relaxable

.option norelax
j label   // redundant relocation due to ForceRelocs
.option relax

label:
```

Root problem: The `isSymbolRefDifferenceFullyResolvedImpl` condition in
MCAssembler::evaluateFixup does not check whether two locations are
separated by a fragment whose size can be indeterminate due to linker
instruction (e.g. MCDataFragment with relaxation, or MCAlignFragment
due to indeterminate start offst).

This patch

* Updates the fragment walk code in
  `attemptToFoldSymbolOffsetDifference` to treat MCRelaxableFragment
  (for --riscv-asm-relax-branches) as fixed size after finishLayout.
* Adds a condition in `addReloc` to complement
  `isSymbolRefDifferenceFullyResolvedImpl`.
* Removes the no longer needed `shouldForceRelocation`.

This fragment walk code path handles nicely handles
mixed relax/norelax case from
https://discourse.llvm.org/t/possible-problem-related-to-subtarget-usage/75283
and allows us to remove `MCSubtargetInfo` argument (#73721) as a follow-up.

This fragment walk code should be avoided in the absence of
linker-relaxable fragments within the current section.

Adjust two bolt/test/RISCV tests (#141310)

MaskRay committed:
[𝘀𝗽𝗿] initial version

Created using spr 1.3.5-bogner

MaskRay committed:
better comments

Created using spr 1.3.5-bogner

MaskRay committed:
make the condition more aggressive

Created using spr 1.3.5-bogner

MaskRay committed:
simplify after 6f6dc1f239433393c0b09430193beb85d03464c8

Created using spr 1.3.5-bogner

MaskRay commented:
Ready for review:) This cleans up tech debt in our relocation generation, which I hope we get in before introducing more stuff to `addReloc`.

I will do a rebase soon to re-trigger CI.

MaskRay committed:
rebase after 68472a39a0fbf38f5da7bb4ebe43e2ca87ff8308

Created using spr 1.3.5-bogner

MaskRay committed:
update objdump test

Created using spr 1.3.5-bogner

lenary reviewed (approved):
LGTM. Thanks for this cleanup, this seems more obvious without the setForceRelocs() calls

MaskRay committed:
remove redundant relocations for pure no-relax code

Created using spr 1.3.5-bogner

MaskRay commented:
* Renamed `MCFixup::NeedsRelax` to `LinkerRelaxable` (which populates the parent MCFragment::LinkerRelaxable). I don't remember why I picked `NeedsRelax`, but since we have both assembler relaxation and linker relaxation, we need a clearer name.
* Introduced `MCSection::LinkerRelaxable` to prevent pessimization when linker relaxation is disabled. We need a reliably way to prevent relocations within a section for no-relax code.

lenary reviewed (approved):
I agree these new names are better for distinguishing branch vs linker relaxation. One nit, but you can skip it if you want.

MaskRay commented on llvm/lib/MC/MCELFStreamer.cpp:
```diff
@@ -450,8 +450,10 @@ void MCELFStreamer::emitInstToData(const MCInst &Inst,
     Fixup.setOffset(Fixup.getOffset() + CodeOffset);
-    if (Fixup.needsRelax())
+    if (Fixup.isLinkerRelaxable()) {
       DF->setLinkerRelaxable();
```
MCFragment::Parent can be non-null (the fragment is in a section) or null (newly created fragment, not added to a section yet).

While DF->setLinkerRelaxable could set its Parent's LinkerRelaxable, it would require MCFragment to know MCSection, which I want to avoid...

MaskRay committed:
adjust bolt tests

Created using spr 1.3.5-bogner
